src/semaphore.rs

```diff
@@ -70,41 +121,131 @@ impl Semaphore {
+    /// The layouts given with each texture must match the image layout used in Vulkan directly before the semaphore is signalled by it.
+    pub fn wait<T: ?Sized>(
+        &self,
+        textures: Option<&[(TextureAny, TextureLayout)]>,
+        buffers: Option<&[Buffer<T>]>,
+    ) where
+        T: Content,
+    {
         let ctxt = self.context.get_context().make_current();
+
+        let (buffer_ids, buffer_num, _) = match buffers {
+            Some(buffs) => {
+                let ids = buffs.iter().map(|b| b.get_id()).collect::<Vec<_>>();
+                (ids.as_ptr(), buffs.len(), Some(ids))
+            }
+            None => (std::ptr::null(), 0, None),
+        };
+
+        let (texture_ids, texture_layouts, textures_num, _, _) = match textures {
+            Some(textures) => {
+                let ids = textures.iter().map(|t| t.0.get_id()).collect::<Vec<_>>();
+                let layouts = textures
+                    .iter()
+                    .map(|t| t.1.into())
+                    .collect::<Vec<gl::types::GLenum>>();
+                (
+                    ids.as_ptr(),
+                    layouts.as_ptr(),
+                    textures.len(),
+                    Some(ids),
+                    Some(layouts),
+                )
+            }
+            None => (std::ptr::null(), std::ptr::null(), 0, None, None),
+        };
+
         unsafe {
             ctxt.gl.WaitSemaphoreEXT(
                 self.id,
-                0,
-                std::ptr::null(),
-                0,
-                std::ptr::null(),
-                std::ptr::null(),
+                buffer_num as u32,
+                buffer_ids,
+                textures_num as u32,
+                texture_ids,
+                texture_layouts,
             )
         }
     }
 
+    /// Same as `signal`, but without buffers parameter
+    pub fn signal_textures(&self, textures: Option<&[(TextureAny, TextureLayout)]>) {
+        // Don't care about type parameter.
+        self.signal::<u32>(textures, None)
+    }
+
     /// Sends signal through semaphore.
-    pub fn signal(&self) {
+    /// The memory of the `textures` and `buffers` is made available to the external API.
+    /// Before signalling the semaphore, the `textures` are transitioned into the layout specified.
+    pub fn signal<T: ?Sized>(
+        &self,
+        textures: Option<&[(TextureAny, TextureLayout)]>,
+        buffers: Option<&[Buffer<T>]>,
+    ) where
+        T: Content,
+    {
+        let (buffer_ids, buffer_num, _) = match buffers {
+            Some(buffs) => {
+                let ids = buffs.iter().map(|b| b.get_id()).collect::<Vec<_>>();
+                (ids.as_ptr(), buffs.len(), Some(ids))
+            }
+            None => (std::ptr::null(), 0, None),
+        };
+
+        let (texture_ids, texture_layouts, textures_num, _, _) = match textures {
+            Some(textures) => {
```

Comment:
This should be an `if let`

Reply:
I think in this case the argument could be made that `match` is the better choice, since you are indeed matching exhaustively, and in my opinion in this particular case it is clearer.  Is there some reasoning that I am missing which makes `if let` the superior option here?

Also, have you considered adding clippy to the pipeline in order to catch this type of mistakes? (It doesn't seem to give this recommendation)

Reply:
`match` adds two indentation levels while `if let` only adds one.

I'm not a fan of all clippy lints but this suggestion by clippy is good.